ktor-server/ktor-server-core/jvm/src/io/ktor/response/ApplicationResponseFunctions.kt

```diff
@@ -12,22 +12,29 @@ import io.ktor.http.content.*
 import io.ktor.util.*
 import io.ktor.utils.io.*
 import java.io.*
+import kotlin.reflect.*
 
 /**
  * Sends a [message] as a response
  */
-@Suppress("NOTHING_TO_INLINE")
-public suspend inline fun ApplicationCall.respond(message: Any) {
-    response.pipeline.execute(this, message)
+@OptIn(ExperimentalStdlibApi::class)
+public suspend inline fun <reified T> ApplicationCall.respond(message: T) {
+    response.pipeline.intercept(ApplicationSendPipeline.Before) {
+        call.attributes.put(ResponseTypeAttributeKey, typeOf<T>())
+    }
+    response.pipeline.execute(this, message as Any)
 }
 
 /**
  * Sets [status] and sends a [message] as a response
  */
-@Suppress("NOTHING_TO_INLINE")
-public suspend inline fun ApplicationCall.respond(status: HttpStatusCode, message: Any) {
+@OptIn(ExperimentalStdlibApi::class)
+public suspend inline fun <reified T> ApplicationCall.respond(status: HttpStatusCode, message: T) {
```

Comment:
Intercepting a call pipeline is a relatively expensive operation that introduce a lot of allocations during pipeine merge. Also, it looks like we are adding this type every time even if we don't need it so we pay even with no serialization enabled

Reply:
removed interception. unfortunately, we can't remove the evaluation of type here, because it's the only place where we still can access it